lib/active_model_serializers/adapter/json_api/deserialization.rb

```diff
@@ -188,7 +188,9 @@ def parse_relationship(assoc_name, assoc_data, options)
             end
 
           polymorphic = (options[:polymorphic] || []).include?(assoc_name.to_sym)
-          hash["#{prefix_key}_type".to_sym] = assoc_data[:type] if polymorphic
+          if polymorphic
+            hash["#{prefix_key}_type".to_sym] = assoc_data.present? ? assoc_data[:type] : nil
```

Comment:
I did try polymorphic && assoc_data earlier, it ended up giving me just the id field and not the type field.